src/main/java/com/microsoft/aad/adal4j/SSLProxyTunnelSocketFactory.java

```diff
@@ -0,0 +1,159 @@
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package com.microsoft.aad.adal4j;
+
+import org.apache.commons.codec.binary.Base64;
+
+import javax.net.ssl.SSLSocket;
+import javax.net.ssl.SSLSocketFactory;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.PrintStream;
+import java.net.InetAddress;
+import java.net.Socket;
+import java.util.Scanner;
+
+public class SSLProxyTunnelSocketFactory extends SSLSocketFactory {
+    private SSLSocketFactory defaultFactory;
+
+    private String tunnelHost;
+
+    private int tunnelPort;
+
+    private String proxyUserName;
+
+    private String proxyPassword;
+
+    public SSLProxyTunnelSocketFactory(String proxyHost, int proxyPort,
+                                       String proxyUserName, String proxyPassword) {
+        tunnelHost = proxyHost;
+        tunnelPort = proxyPort;
+        defaultFactory = (SSLSocketFactory) SSLSocketFactory.getDefault();
+
+        this.proxyUserName = proxyUserName;
+        this.proxyPassword = proxyPassword;
+    }
+
+    public Socket createSocket(String host, int port) throws IOException {
+        return createSocket(null, host, port, true);
+    }
+
+    public Socket createSocket(String host, int port, InetAddress clientHost,
+                               int clientPort) throws IOException {
+        return createSocket(null, host, port, true);
+    }
+
+    public Socket createSocket(InetAddress host, int port) throws IOException {
+        return createSocket(null, host.getHostName(), port, true);
+    }
+
+    public Socket createSocket(InetAddress address, int port,
+                               InetAddress clientAddress, int clientPort) throws IOException {
+        return createSocket(null, address.getHostName(), port, true);
+    }
+
+    public Socket createSocket(Socket s, String host, int port,
```

Comment:
Socket s is passed in but never used. Perhaps if s is not null, tunnel should be set to s ?